Veritas toolkit migration

timlrx opened the pull request:
## Description

Port over [Veritas toolkit](https://github.com/mas-veritas2/veritastool) as an AIV stock plugin. This PR focuses on refactoring the algorithm library to produce an AIV-compatible test output file (compatible with the AIV schema), while preserving as much backward compatibility for existing Veritas users.

There are two ways to use the plugin:
1. As part of the data discovery and modelling process within a notebook environment, similar to existing Veritas usage - see example notebooks.
2. To audit existing models against a specified input parameters via the CLI i.e. the AI Verify way.

The output format has now been aligned with the AIVerify output schema. To port over existing code, there is a `convert_veritas_artifact_to_aiverify` util function that can be used to convert previous Veritas output to the AIVerify format. Example usage:

```py
from aiverify_veritastool.util.aiverify import convert_veritas_artifact_to_aiverify

convert_veritas_artifact_to_aiverify(model_artifact_path="veritas_output.json", output_dir="output")
```

Porting of front-end functionality as AIV widgets will be handled subsequently.

## Tasks

- [x]  Migrate veritas codebase to AIV
- [x]  Updated veritas dependencies, data and models
- [x]  Add aiverify-test-engine support for reading and serializing pipeline files used in Veritas 
- [x] Align AIV output to Veritas format
- [x] Support BaseClassification
- [x] Support BaseRegression
- [x] Support CreditScoring
- [x] Support CustomerMarketing
- [x] Support PredictiveUnderwriting
- [x] Add input schema checks
- [x] Add output schema checks
- [x] Add wrapper to convert from Veritas output to AIV output

## Type of Change

<!-- Select the appropriate type of change: -->
<!-- - feat: A new feature -->
<!-- - chore: Routine tasks, maintenance, or tooling changes -->
<!-- - docs: Documentation updates -->
<!-- - refactor: Code refactoring without changes in functionality -->

## How to Test

Unit test - `python -m pytest tests/util tests/metrics tests/models tests/principles tests/usecases -p no:warnings`

Note: The util test is a little flaky and has to be run first
![CleanShot 2025-01-20 at 18 39 38](https://github.com/user-attachments/assets/7d79a6dc-fab9-471a-8d84-3c456e190dbf)

E2E test i.e. running the Veritas plugin the AIVerify way - `python -m pytest tests/e2e -p no:warnings`:

![CleanShot 2025-01-21 at 17 07 05](https://github.com/user-attachments/assets/21c86a1e-0628-4d88-8bcb-67385a610c9c)

Alternatively, you could modify the root path accordingly and run the following bash script:

```sh
root_path="/Users/Documents/aiverify/stock-plugins/user_defined_files/veritas_data"
python "aiverify.stock.veritas/algorithms/veritastool/aiverify_veritastool/plugin_init.py" \
  --data_path $root_path/cs_X_test.pkl \
  --model_path $root_path/cs_model.pkl \
  --ground_truth_path $root_path/cs_y_test.pkl \
  --ground_truth y_test \
  --training_data_path $root_path/cs_X_train.pkl \
  --training_ground_truth_path $root_path/cs_y_train.pkl \
  --training_ground_truth y_train \
  --use_case "base_regression" \
  --privileged_groups '{"SEX": [1], "MARRIAGE": [1]}' \
  --model_type CLASSIFICATION \
  --fair_threshold 80 \
  --fair_metric "auto" \
  --fair_concern "eligible" \
  --performance_metric "accuracy" \
  --transparency_rows 20 40 \
  --transparency_max_samples 1000 \
  --transparency_features LIMIT_BAL \
  --run_pipeline
```

Integration test with API-GW:

Tested on API-GW with docker image and added the veritas plugin as a custom plugin before uploading the output from the 5 example test cases that corresponds with the jupyter notebook example output

![CleanShot 2025-01-20 at 18 14 29](https://github.com/user-attachments/assets/db403193-35d2-4416-880d-8b3599dbb1ad)
![CleanShot 2025-01-20 at 18 15 21](https://github.com/user-attachments/assets/50601ea2-960b-40ba-9d08-1436e486e542)
![CleanShot 2025-01-20 at 18 15 44](https://github.com
![CleanShot 2025-01-20 at 18 16 01](https://github.com/user-attachments/assets/a261eb42-9c04-4d08-812c-2ea8ecd29ade)
/user-attachments/assets/8dfc71be-13f6-4e48-a312-1aea3562714d)
![CleanShot 2025-01-20 at 18 16 20](https://github.com/user-attachments/assets/aa38b6b4-ffdb-4a79-bd78-be14fe705c29)


## Checklist

Please check all the boxes that apply to this pull request using "x":

- [x]  I have tested the changes locally and verified that they work as expected.
- [x]  I have added or updated the necessary documentation (README, API docs, etc.).
- [x]  I have added appropriate unit tests or functional tests for the changes made.
- [x]  I have followed the project's coding conventions and style guidelines.
- [x]  I have rebased my branch onto the latest commit of the main branch.
- [x]  I have squashed or reorganized my commits into logical units.
- [x]  I have added any necessary dependencies or packages to the project's build configuration.
- [x]  I have performed a self-review of my own code.
- [x]  I have read, understood and agree to the Developer Certificate of Origin below, which this project utilises.

## Additional Notes

[Add any additional information or context that might be relevant to reviewers.]

<details>
  <summary>Developer Certificate of Origin</summary>

 ```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
 ```
</details>

imda-peckyoke commented:
@timlrx will it be better to specify the dependencies for aiverify-test-engine to be "aiverify-test-engine===2.*" instead of "aiverify-test-engine==2.0.0a2". Right now if i just run pip install without manuallly install the aiverify-test-engine first before hand, I'll get an error because 2.0.0a2 not published yet

timlrx commented:
> @timlrx will it be better to specify the dependencies for aiverify-test-engine to be "aiverify-test-engine==2.*" instead of "aiverify-test-engine==2.0.0a2". Right now if i just run pip install without manuallly install the aiverify-test-engine first before hand, I'll get an error because 2.0.0a2 not published yet

Hi @imda-peckyoke, for this stock-plugin we need to use the latest version so I just specified 2.0.0a2 in advance. Perhaps we can add it as `aiverify-test-engine==^2.0.0a2` so newer versions are also supported?

imda-peckyoke commented:
@timlrx can you change the gid in algo.meta.json to "aiverify.stock.veritas"? It must match the gid in plugin.meta.json.

imda-peckyoke commented:
The size of the veritas output results.json is too big (~50mb), which exceeds the size limit of multipart/form-data. Cannot use "/test_results/upload" to upload the results to API-GW; need to zip the output and upload using "/test_results/upload_zip" endpoint instead. @imda-benedictlee please note for your integration tests. Thanks

imda-peckyoke commented:
@timlrx are images generated in the output meant to be uploaded to the apigw? If so, you need to update the results.json to include an `artifacts` field specifying list of the artifacts to save.

timlrx commented:
@imda-peckyoke, thanks for catching that. Will make the necessary changes.

timlrx commented:
Fixed the 3 issues raised above:
1. Changed dependency to `aiverify-test-engine==^2.0.0a2`
2. Updated gid in `algo.meta.json` to "aiverify.stock.veritas"
3. Update `results.json` to include an artifacts field specifying the list of artifacts to save

Added a test case for item 3. as well:
![CleanShot 2025-01-24 at 13 39 28](https://github.com/user-attachments/assets/add13110-e5bb-4304-960d-484af3756a1e)

Also attached a sample regression output from the updated code, that you could use for testing purposes:
[output.zip](https://github.com/user-attachments/files/18531128/output.zip)

imda-peckyoke commented:
@timlrx the artifacts is still not in results.json

timlrx commented:
> @timlrx the artifacts is still not in results.json

Pushed the fix.

timlrx commented:
Added dockerfile.

imda-peckyoke commented:
@timlrx one more bug. When I use convert_veritas_artifact_to_aiverify to covert the output from the original veritas notebook to AIVT2 output format, the `artifacts` list  is under the `output` field. The `artifacts` field should be under the parent (root) object instead.  See attached.

[model_artifact_credit_scoring_20250124_1429.json](https://github.com/user-attachments/files/18556570/model_artifact_credit_scoring_20250124_1429.json)

[CS_demo_results.zip](https://github.com/user-attachments/files/18556579/CS_demo_results.zip)

timlrx commented:
@imda-peckyoke, pushed the fix. The `artifacts` field is now at the root level.

imda-peckyoke reviewed (approved):
Approved.
